server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java

```diff
@@ -382,6 +389,38 @@ public String get(String key) {
         }
     }
 
+    public static String getProviderProfile(final Scope originalScope, final String spiName, final String provider) {
+
+        // check if the provider profile was configured in the original scope.
+        String profile = originalScope.get("provider-profile");
+        if (profile == null || profile.isEmpty()) {
+            // profile not configured, find the default profile in the root provider configuration.
+            Scope providerScope = Config.scope(spiName, provider);
+            String defaultProfile = providerScope.get("default-profile");
+            if (defaultProfile != null && !defaultProfile.isEmpty()) {
+                // default profile was set in the root provider configuration.
+                profile = defaultProfile;
+            } else {
+                profile = "default";
+            }
+        }
+        return profile;
+    }
+
+    public static Scope getProviderScope(final Scope rootScope, final String profile) {
+        if (profile == null) {
+            return rootScope;
+        }
+        else if (profile.equals("default")) {
+            // we must either have a profile named default, or we are using the provider root config.
+            Scope defaultProfileScope = rootScope.scope("profile", "default");
+            return defaultProfileScope.isEmpty() ? rootScope : defaultProfileScope;
+        } else {
+            // non-default profile, adjust the scope to the specified profile.
+            return rootScope.scope("profile", profile);
+        }
+    }
```

Comment:
After I've finished implementing and testing this, I realized I can probably do without the "profile" scope altogether. So, instead of a config like this:

```
"mapStorage": {
    "jpa-map-storage": {
        "profile": {
             "xyz": {
             }
        }
    }
}
```
I could probably have just
```
"mapStorage": {
    "jpa-map-storage": {
         "xyz": {
         }
    }
}
```

Reply:
I think this is not a good way forward - the default configuration keys would be indistinguishable and might interfere with profile name.

Reply:
Good point, will stick with explicit `profile` scope scope.